### PR TITLE
Netskope event collector rewrite

### PR DESCRIPTION
**Pull request by ShahafBenYakir**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Jira ticket
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-7622

## Description
This is the mew implementation of Netskope event collector using the v2 dataexport

### CODE REVIEW

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml [hunk 1]
```diff
@@ -17,28 +17,19 @@ configuration:
...
   name: first_fetch
-  section: Collect
   type: 0
+  section: Collect
   required: false
-- additionalinfo: The maximum amount of events to retrieve for each event type (up to 30000 events). For more information about event types see the help section.
-  defaultvalue: '1000'
+  hidden: true
+- additionalinfo: 'The maximum amount of events to retrieve PER EACH EVENT TYPE type. For more information about event types see the help section.'
+  defaultvalue: '50000'
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml:29]: Don't you prefer to lower the default value?

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml [hunk 2]
```diff
@@ -17,28 +17,19 @@ configuration:
...
-  defaultvalue: v2
-  display: API Version
-  name: api_version
-  type: 15
-  section: Connect
-  options:
-  - v1
-  - v2
-  required: false
 - additionalinfo: 'The first fetch is limited to the last 50,000 events (for each event type).'
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml:20]: I don't think this is necessary anymore

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py [hunk 3]
```diff
@@ -29,203 +35,249 @@ class Client(BaseClient):
...
+    def __init__(self, base_url: str, token: str, validate_certificate: bool, proxy: bool):
+        self.fetch_status: dict = {event_type: False for event_type in ALL_SUPPORTED_EVENT_TYPES}
 
-    def get_events_request_v2(self, event_type: str, last_run: dict, skip: int = None,
-                              limit: int = None, is_command: bool = False) -> Dict:  # pragma: no cover
+        headers = {'Netskope-Api-Token': token}
+        super().__init__(base_url, verify=validate_certificate, proxy=proxy, headers=headers)
 
-        url_suffix = f'events/data/{event_type}'
+    def perform_data_export(self, endpoint, _type, index_name, operation):
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py:44]: Let's add type annotations

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py [hunk 4]
```diff
@@ -244,70 +296,65 @@ def get_events_command(client: Client, args: Dict[str, Any], last_run: dict, api
...
-        last_run = demisto.getLastRun()
+        params = demisto.params()
+
+        url = params.get('url')
+        token = params.get('credentials', {}).get('password')
+        base_url = urljoin(url, '/api/v2/')
+        verify_certificate = not params.get('insecure', False)
+        proxy = params.get('proxy', False)
+        max_fetch = arg_to_number(params.get('max_fetch', 10000))
+        vendor, product = params.get('vendor', 'netskope'), params.get('product', 'netskope_dev')
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py:312]:
```suggestion
        vendor, product = params.get('vendor', 'netskope'), params.get('product', 'netskope')
```

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py [hunk 5]
```diff
@@ -29,203 +35,249 @@ class Client(BaseClient):
...
 
-    Returns (list): list of all the events from a start time.
+    Returns:
+        dict: the modified last run dictionary with the needed operation
     """
-    events_result = []
-    new_last_run = {}
-    if limit is None:
-        limit = MAX_EVENTS_PAGE_SIZE
+    first_fetch = int(arg_to_datetime('now').timestamp())  # type: ignore[union-attr]
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py:163]: So the first_fetch will always be from now?
We won't support going few days back?

File: Packs/Netskope/Integrations/NetskopeEventCollector/README.md [hunk 6]
```diff
@@ -8,29 +8,31 @@
     | --- | --- | --- |
     | Server URL |  | True |
     | API token |  | True |
-    | API Version | The API version to use \(v1 or v2\). | False |
     | Trust any certificate (not secure) |  | False |
     | Use system proxy settings |  | False |
-    | First fetch timestamp (&lt;number&gt; &lt;time unit&gt;, e.g., 12 hours, 7 days) |  | False |
-    | Max events per fetch | The maximum amount of events to retrieve \(up to 30000 events\). | False |
+    | Max events per fetch | The maximum amount of events to retrieve per each event type. For more information about event types see the help section. | False |
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/README.md:13]: Don't you want to add the limitations also to the readme? 
Under the known limitation section?

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector_test.py [hunk 7]
```diff
@@ -1,46 +1,29 @@
...
         return json.loads(f.read())
 
 
-MOCK_ENTRY = util_load_json('test_data/mock_events.json')
-EVENTS_RAW_V2 = util_load_json('test_data/events_raw_v2.json')
-EVENTS_RAW_V2_MULTI = util_load_json('test_data/events_raw_v2_2_results.json')
-EVENTS_PAGE_RAW_V1 = util_load_json('test_data/page_raw_v1.json')
+MOCK_ENTRY = util_load_json('test_data/mock_events_entry.json')
+EVENTS_RAW = util_load_json('test_data/events_raw.json')
+# EVENTS_RAW_V2_MULTI = util_load_json('test_data/events_raw_v2_2_results.json')
```
Comment (Shellyber) [inline Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector_test.py:19]: If not needed. please remove

**Review by Shellyber — COMMENTED** [review]
Great Work!

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml (hunk 1 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml:29]: Why actually?

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py (hunk 4 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py:312]: good catch

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py (hunk 5 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py:163]: Nope, per new standards for collectors design we don't need to support fetching backwards unless specifically asked

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml (hunk 1 above)
Comment (Shellyber) [reply Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml:29]: To not take risks, not to overwhelm the system too much if it is not necessary.
No reason not to start with a bit lower value and if the client wants, he can increase (30K for example) but whatever you think

File: Packs/Netskope/Integrations/NetskopeEventCollector/README.md (hunk 6 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/README.md:13]: WDYM? I have this https://github.com/demisto/content/blob/c9a58b26fef538397c2572b9836c04ba1f719846/Packs/Netskope/Integrations/NetskopeEventCollector/README.md#fetch-events-limitation

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml (hunk 1 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml:29]: Don't see a reason we can digest 50k with no problem. We also have the timeout mechanism for extra safety

File: Packs/Netskope/Integrations/NetskopeEventCollector/README.md (hunk 6 above)
Comment (Shellyber) [reply Packs/Netskope/Integrations/NetskopeEventCollector/README.md:13]: My bad - Missed that.
Do we want to add also the API limitation regarding indexation?

File: Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml (hunk 2 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.yml:20]: The entire arg is actually not needed anymore

File: Packs/Netskope/Integrations/NetskopeEventCollector/README.md (hunk 6 above)
Comment (ShahafBenYakir) [reply Packs/Netskope/Integrations/NetskopeEventCollector/README.md:13]: Waiting for the Netskope team investigation resolution on this

**Review by Shellyber — APPROVED** [review]

**Comment by ShahafBenYakir** [comment]
Validation fails on removed parameters, this is by design as the first fetch is not relevant anymore

**Pull request merged**